Add Token.parse and Token.parseMultiple to parse response token

thibmeu opened the pull request:
Challenge headers can be parsed via `PrivateToken.parse`. Response tokens should have the same functionality, as they would need to be parsed by attester/issuer, and serialised by origins and clients.

This commit adds this features, along with associated test over provided test vectors.

cdrubin commented on test/pubVerifToken.test.ts:
```diff
@@ -83,4 +83,14 @@ test.each(vectors)('PublicVerifiable-Vector-%#', async (v: Vectors) => {
 
     expect(await token.verify(publicKey)).toBe(true);
+
+    const header = token.toString();
```
I thought of suggesting a test of a header containing multiple tokens or additional attributes but in my test these parsed nicely anyway.

armfazh replied:
agree, more coverage is needed, e.g. parsing for multiple tokens  or a mix of valid and invalid tokens. Likely in other PR, to unblock this one.

cdrubin reviewed (approved):
This looks reasonable and works correctly for me.

armfazh commented on src/httpAuthScheme.ts:
```diff
@@ -370,6 +370,75 @@ export class Token {
     }
 
+    static parse(tokenTypeEntry: TokenTypeEntry, data: string): Token {
+        // Consumes data:
+        //   challenge="abc...", token-key="123..."
```
not sure the docstring is correct.

thibmeu replied:
updated

armfazh commented on src/httpAuthScheme.ts:
```diff
@@ -370,6 +370,75 @@ export class Token {
+                case 'token': {
+                    const tokenEnc = base64url.parse(attrValue);
+                    ppToken = Token.deserialize(tokenTypeEntry, tokenEnc);
+                    break;
```
breaks the switch-case or breaks the for-loop?
returns always the first one?

thibmeu replied:
I removed the switch, but kept the for loop. WWW-Authenticate header can contain more than one attribute, even if not specified by the protocol.

armfazh commented on test/pubVerifToken.test.ts:
```diff
@@ -83,4 +83,14 @@ test.each(vectors)('PublicVerifiable-Vector-%#', async (v: Vectors) => {
+    expect(parsedToken.payload.nonce).toEqual(token.payload.nonce);
+    expect(parsedToken.payload.tokenKeyId).toEqual(token.payload.tokenKeyId);
+    expect(parsedToken.payload.tokenType).toBe(token.payload.tokenType);
+    expect(parsedToken.authenticator).toEqual(token.authenticator);
```
it would be good to have in the reference implementation the test vector for token

```diff
{
   "WWW-Authenticate": "PrivateToken challenge=abc...",
+  "Authorization": "PrivateToken token=abc..."
}
```

armfazh commented on src/httpAuthScheme.ts:
```diff
@@ -370,6 +370,75 @@ export class Token {
+        // WWW-Authenticate does not impose authentication parameters escape with a double quote
+        // For more details, refer to RFC9110 Section 11.2 https://www.rfc-editor.org/rfc/rfc9110#section-11.2
+        const quote = quotedString ? '"' : '';
+        // eslint-disable-next-line func-style
+        const authParamToString = (param: string, value: string | number): string =>
+            `${param}=${quote}${value}${quote}`;
+        const token = base64url.stringify(this.serialize());
+        const authParams: Record<string, string | number> = {
+            token,
+        };
+        const params = Object.entries(authParams)
+            .map(([param, value]) => authParamToString(param, value))
+            .join(', ');
+        return `PrivateToken ${params}`;
```
it is simpler to do:

```suggestion
        const token = base64url.stringify(this.serialize());
        return `PrivateToken token=${quote}${token}${quote}`;
```

thibmeu replied:
I moved the `toString` function to rfc9110.ts, and updated both Token and PrivateToken methods to use it.

armfazh reviewed (approved):
few minor things

thibmeu committed:
Add Token.parse and Token.parseMultiple to parse response token

Challenge headers can be parsed via `PrivateToken.parse`. Response
tokens should have the same functionality, as they would need to be
parsed by attester/issuer, and serialised by origins and clients.

This commit adds this features, along with associated test over provided
test vectors.
